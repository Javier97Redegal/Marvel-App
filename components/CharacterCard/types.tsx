type CharacterCardProps = {
    character: {
        id: number
        name: string | undefined
        thumbnail: {
            path: string
            extension: string
        }
    }
}