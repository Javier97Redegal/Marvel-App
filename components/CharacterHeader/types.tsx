export interface CharacterHeaderProps {
    id: number
}