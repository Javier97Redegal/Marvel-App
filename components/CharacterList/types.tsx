export interface CharacterListProps {
    mode: string
    query: string
}