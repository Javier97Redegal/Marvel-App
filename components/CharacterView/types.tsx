export interface CharacterViewProps {
    id: string
}