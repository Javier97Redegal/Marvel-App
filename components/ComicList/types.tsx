export interface ComicListProps {
    id: number
}