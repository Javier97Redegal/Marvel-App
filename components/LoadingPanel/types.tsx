export interface LoadingPanelProps {
    type: 'error' | 'loading' | 'idle' | 'success'
}