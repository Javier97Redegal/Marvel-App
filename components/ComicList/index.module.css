.comics {
    max-width: 1980px;
    padding: 2rem;
    margin: 0 auto;
}

.comics ul {
    display: flex;
    gap: 1rem;
    padding: 0;
    list-style: none;
    overflow-x: scroll;
}

.comics ul li {
    width: 20%;
    max-width: 250px;
    flex: 0 0 auto;
}

.header {
    font-size: 2rem;
    font-weight: 700;
    text-transform: uppercase;
}