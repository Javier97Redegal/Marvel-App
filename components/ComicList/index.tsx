import { useQuery } from 'react-query'
import { getComicsFromCharacter } from '@/utils/MarvelAPI'
import ComicCard from '@/components/ComicCard'
import LoadingPanel from '@/components/LoadingPanel'
import styles from './index.module.css'
import { ComicListProps } from './types'
import { DEFAULT_COMIC_LIMIT } from '@/constants'

const ComicList = ({
    id
}: ComicListProps) => {
    const fetchComics = async () => await getComicsFromCharacter({ id, params: { limit: DEFAULT_COMIC_LIMIT } })
    const { data, status } = useQuery(`comics-${id}`, fetchComics)

    if (status !== 'success') return <LoadingPanel type={status} />

    const comics = data.data.results

    return <>
        <div className={styles.comics}>
            <h2 className={styles.header}>Comics</h2>
            {!!comics.length ?
                <ul>
                    {comics.map(comic => (
                        <li key={comic.id}>
                            <ComicCard {...comic}></ComicCard>
                        </li>
                    ))}
                </ul> :
                <LoadingPanel type='success' />
            }
        </div>
    </>
}

export default ComicList