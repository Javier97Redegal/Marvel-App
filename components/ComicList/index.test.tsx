import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { useQuery } from 'react-query'
import ComicList from '.'
import { ComicListProps } from './types'

jest.mock('react-query', () => ({
    useQuery: jest.fn()
}))

jest.mock('@/utils/MarvelAPI', () => ({
    getComicsFromCharacter: jest.fn()
}))

jest.mock('@/components/ComicCard', () => jest.fn(({ title }: { title: string }) => <div>{title}</div>))
jest.mock('@/components/LoadingPanel', () => jest.fn(({ type }: { type: string }) => <div>LoadingPanel Mock: {type}</div>))

describe('ComicList tests', () => {
    const mockProps: ComicListProps = { id: 12345 }
    const comicsData = {
        data: {
            results: [
                { id: '1', title: 'Comic 1' },
                { id: '2', title: 'Comic 2' }
            ]
        }
    }

    it('renders loading panel when loading', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'loading' })

        const { asFragment } = render(<ComicList {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: loading')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders loading panel when error', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'error' })

        const { asFragment } = render(<ComicList {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: error')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders list of comics when data is loaded', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: comicsData, status: 'success' })

        const { asFragment } = render(<ComicList {...mockProps} />)

        expect(screen.getByText('Comic 1')).toBeInTheDocument()
        expect(screen.getByText('Comic 2')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders loading panel when there are no comics', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: { data: { results: [] } }, status: 'success' })

        const { asFragment } = render(<ComicList {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: success')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })
})
