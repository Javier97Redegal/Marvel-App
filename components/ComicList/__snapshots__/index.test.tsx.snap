// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicList tests renders list of comics when data is loaded 1`] = `
<DocumentFragment>
  <div
    class="comics"
  >
    <h2
      class="header"
    >
      Comics
    </h2>
    <ul>
      <li>
        <div>
          Comic 1
        </div>
      </li>
      <li>
        <div>
          Comic 2
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ComicList tests renders loading panel when error 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: error
  </div>
</DocumentFragment>
`;

exports[`ComicList tests renders loading panel when loading 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: loading
  </div>
</DocumentFragment>
`;

exports[`ComicList tests renders loading panel when there are no comics 1`] = `
<DocumentFragment>
  <div
    class="comics"
  >
    <h2
      class="header"
    >
      Comics
    </h2>
    <div>
      LoadingPanel Mock: success
    </div>
  </div>
</DocumentFragment>
`;
